### ISSUE TITLE
Submission: househunter (R)

### ISSUE THREAD

**Issue by elina-linglin** [UBC-MDS/software-review-2021#23]
name: R House Hunter
Submitting Author: Ela Bandari (@elabandari ), 
&nbsp;   &nbsp;  &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;  &nbsp;Junting He (@JuntingHe ),
&nbsp;   &nbsp;  &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;  &nbsp;Elina (Ling) Lin (@elina-linglin ),   
&nbsp;   &nbsp;  &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;   &nbsp;  &nbsp;Alex Truong (@athy9193)
Repository: https://github.com/UBC-MDS/rhousehunter
Version submitted: 0.1.0
Editor: Tiffany Timbers (@ttimbers )
Reviewer 1: TBD
Reviewer 2: TBD
Archive: TBD
Version accepted: TBD

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: rhousehunter
Title: Simplify House-Hunting
Version: 0.0.0.9000
Authors@R: c(person(given = "Ela", 
                    family = "Bandari",
                    role = "aut"),
              person(given = "Junting", 
                     family = "He",
                     role = "aut"),
              person(given = "Elina", 
                     family = "Lin",
                     role = c("aut", "cre"),
                     email = "elina702@hotmail.com"),
              person(given = "Alex", 
                     family = "Truong",
                     role = "aut"))
Description: This R package includes four functions that simplify rental searching on Craigslist: scrape rental websites on Craiglist, clean the data, filter the data by user inputs, and email users with the organized information.
License: MIT + file LICENSE
Encoding: UTF-8
LazyData: true
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.1.1
URL: https://github.com/UBC-MDS/rhousehunter
BugReports: https://github.com/UBC-MDS/rhousehunter/issues
Config/testthat/edition: 3
Imports: 
    polite,
    rvest,
    tibble,
    xml2,
    testthat,
    dplyr,
    stringr,
    httr,
    readr,
    rlang,
    emayili,
    qpdf
Suggests: 
    covr,
    knitr, 
    rmarkdown
Depends: 
    R (>= 2.10)
VignetteBuilder: knitr 
```

## Scope

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [ ] data retrieval
	- [X] data extraction
	- [X] data munging
	- [ ] data deposition
	- [ ] workflow automation
	- [ ] version control
	- [ ] citation management and bibliometrics
	- [ ] scientific software wrappers
	- [ ] field and lab reproducibility tools
	- [ ] database software bindings
	- [ ] geospatial data
	- [ ] text analysis

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):

> Our package automates the process of rental searching with the following four functionalities: scraping housing listings from Craigslist, cleaning the data, filtering out relevant information based on user inputs, and emailing organized information to the user. 

-   Who is the target audience and what are scientific applications of this package?

> The target audience can be anyone who is looking for rental information in Canada. This package could benefit scientific projects that require a large quantity of housing information for data analysis. 

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

> There exist other R packages that facilitate some but not all steps of the entire house hunting process. To help simplify this whole process, our package provides comprehensive functionality that takes care of all the intermediate steps for the user. 

-   If you made a pre-submission enquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

## Technical checks

Confirm each of the following by checking the box.

- [X] I have read the [guide for authors](https://devguide.ropensci.org/guide-for-authors.html) and [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).

This package:

- [X] does not violate the Terms of Service of any service it interacts with.
- [ ] has a CRAN and OSI accepted license.
- [X] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme).
- [X] includes [documentation with examples for all functions, created with roxygen2](https://ropensci.github.io/dev_guide/building.html#documentation).
- [X] contains a vignette with examples of its essential functions and uses.
- [X] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [X] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage using services such as Travis CI, Coveralls and/or CodeCov.

## Publication options

- [ ] Do you intend for this package to go on CRAN?
- [ ] Do you intend for this package to go on Bioconductor?

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:

<details>
<summary>MEE Options</summary>

- [ ] The package is novel and will be of interest to the broad readership of the journal.
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [X] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.github.io/dev_guide/policies.html#code-of-conduct) during the review process and in maintaining my package should it be accepted.

**Comment by sbabicki** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s)** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions
- [x] **Examples** (that run successfully locally) for all exported functions
- [x] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

>##### For packages co-submitting to JOSS
>
>- [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.theoj.org/about#submission_requirements)
>
>The package contains a `paper.md` matching [JOSS's requirements](https://joss.theoj.org/about#paper_structure) with:
>
>- [ ] **A short summary** describing the high-level functionality of the software
>- [ ] **Authors:**  A list of authors with their affiliations
>- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience.
>- [ ] **References:** with DOIs for all those that have one (e.g. papers, datasets, software).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [x] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package
   and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

Estimated hours spent reviewing: 2 hours

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

Great idea for a package, well done! I was able to successfully install the package with `devtools` by following the usage instructions. All tests passed for my local installation. The vignette was very helpful for walking me through how to use the package. I was impressed to see that every step of the process worked with no issues and it was cool to get an email with my results at the end. Amazing job overall, it was a pleasure reviewing `rhousehunter`.

#### General Suggestions
- In the README the sentence "This Python package intends to facilitate..." should be updated to say R instead of Python.  
- Code coverage could be improved - currently < 85%.
- It would be extra convienient if the output on each line of the vignette was in a different cell so the code can be copied with the "copy to clipboard" button.
- The `polite` package dependency is missing and should be added. 
- There were some suggestions from running `goodpractice::gp()` that could be updated, most were to do with code coverage and more than 80 characters in code lines. 

#### data_cleaner.R
- May be better to import the `stringr` library since it is used so often.
- It would be good to include some error handling for expected data types in addition to confirming the column names match (e.g. check that price is a string; not NA, numeric, or some other unexpected type). 

#### data_filter.R
- Correct typos: "minimun", "number number of bedrooms".
- It would be nice to have some input parameters be optional.
- How do multiple inputs work? E.g. if someone passes in a vector of 10 `area_sqft` how should the function deal with that?

#### polite-scrape.R
- Functions should have `roxygen2` documentation, the source code is a bit difficult to follow without documentation. If the code is not written by your group make sure to use proper citations. 

#### scraper.R
- `polite_download_file` is missing proper `roxygen2` documentation, and the function should be updated or removed before publication. 

#### send_email.R
- Password should not be hardcoded in script as plain text.
- Good tests for failure, but you should also consider including some tests for success.

**Comment by AishwaryaGopal12** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [X] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [X] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
- [X ] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [ X] **Vignette(s)** demonstrating major functionality that runs successfully locally
- [X ] **Function Documentation:** for all exported functions
- [ X] **Examples** (that run successfully locally) for all exported functions
- [ X] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

>##### For packages co-submitting to JOSS
>
>- [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.theoj.org/about#submission_requirements)
>
>The package contains a `paper.md` matching [JOSS's requirements](https://joss.theoj.org/about#paper_structure) with:
>
>- [ ] **A short summary** describing the high-level functionality of the software
>- [ ] **Authors:**  A list of authors with their affiliations
>- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience.
>- [ ] **References:** with DOIs for all those that have one (e.g. papers, datasets, software).

#### Functionality

- [ X] **Installation:** Installation succeeds as documented.
- [X ] **Functionality:** Any functional claims of the software been confirmed.
- [X ] **Performance:** Any performance claims of the software been confirmed.
- [X ] **Automated tests:** Unit tests cover essential functions of the package
   and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [X ] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

Estimated hours spent reviewing: close to 3 hours

- [ X] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

I loved the idea for the package because I went through the hassle of finding a place to rent not too long ago. the installation process was a smooth sail. All in all, amazing job.

Suggestions:
Code coverage percentage can be improved.
There are about 3-4 typos, which is a minor issue.
Some functions do not have the roxygen documentation. It's slightly hard to follow what they do.
Consider moving the hardcoded regex strings to a variable and using the variables in the code to avoid >80 characters warning.

**Comment by athy9193** [comment]
Hi @sbabicki and @AishwaryaGopal12  Thank you very much for your comments. They are most helpful!

Follows are my response to address some of the issues:
- Fixed typo in README 

- Regarding the lack of Roxygen2 documentation for some `scraper()` function, this is because these functions (e.g. `polite_download_file()`) are adopted from `polite` for package developers and not for the end-users of `rhousehunter`. Hence we do not put the function documents in so as to avoid they are being automatically rendered on the Documentation website and confuse end-users. Similarly, we didn't add `polite` in the Dependency section as we use the `polite` package for package developers, not the end-users interface of `polite` package, we have include further clarifications on this in the Credits section.

- On the code coverage percentage, we noticed during testing that the virtual machine on Github action was detected as a robot and not able to access the site despite our function `polite` request setup, which leads to HTTPS 403: Forbidden errors. Meanwhile, we run  `test()` or `check()`  in RStudios, the `scraper()` function works fine. We decided to skip a few tests that involve online interaction so that we can successfully integrate our package with Github Action. Hence, this leads to low test coverage.

Hope that addresses some of your concerns. 

Best,
Alex
